### ISSUE TITLE
Incorrect effect of stability attribute on calling private const fn

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust#75794]
In general we don't require private functions to have stability attributes. Stable functions are free to call private functions, and stable const functions are free to call private const functions that have no stability attribute.

But when a private const function inherits a stability attribute from an enclosing module, it currently can no longer be called from stable const functions. I think this behavior is not correct. For a *private* function, I would think stability attributes should be irrelevant.

This is causing trouble in https://github.com/rust-lang/rust/pull/75772#discussion_r474918624. 

```rust
#![feature(staged_api, const_fn)]
#![stable(feature = "repro", since = "0")]

mod detail {
    #![unstable(feature = "detail", issue = "none")]  // works if removed

    pub(crate) const fn f() {}

    // presumably other public stable things
}

#[stable(feature = "repro", since = "0")]
#[rustc_const_stable(feature = "repro", since = "0")]
pub const fn f() {
    detail::f();
}
```

```console
error[E0723]: can only call other `const fn` within a `const fn`, but `const detail::f` is not stable as `const fn`
  --> src/lib.rs:15:5
   |
15 |     detail::f();
   |     ^^^^^^^^^^^
   |
   = note: see issue #57563 <https://github.com/rust-lang/rust/issues/57563> for more information
   = help: add `#![feature(const_fn)]` to the crate attributes to enable
```

Cross referencing `const fn` tracking issue #57563.

**Comment by oli-obk** [comment]
The reason we have this wanky system is that we didn't use to have `rustc_const_(un)stable` (and we didn't even get both of these at the same time, `rustc_const_stable` came quite a bit later than `rustc_const_unstable`). So yea, this is an emergent system that tried to not have any effect on unstable const fns because there were quite a few of them and annotating them with the fact that their constness is also unstable seemed odd.

I would definitely like to have a hard requirement of either `rustc_const_stable` or `rustc_const_unstable` on all `const fn`, but won't that be counter to the point of being able to quickly evolve an unstable module without requiring all these attributes?

I would have thought that

```rust
#![feature(staged_api, const_fn)]
#![stable(feature = "repro", since = "0")]
#![allow(dead_code)]

#[unstable(feature = "detail", issue = "none")]
pub mod detail {
    #[rustc_const_stable(...)]
    pub(crate) const fn bar() {}
}

// not public, not rustc_const_stable, maybe not even used
const fn foo() {
    detail::bar();
}
```

would compile, which is also how I'd expect all of this to work. Maybe we should not be checking the regular stability of the const fn, but whether the current crate is `staged_api` to decide whether we treat all unmarked const fns as `rustc_const_unstable`?

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-const-eval (looks like the group hasn't been pinged yet)

**Comment by RalfJung** [comment]
@dtolnay isn't it currently the case that all private methods are considered implicitly "unstable" by the stability system? AFAIK in a crate with `staged_api`, *every* function (type, ...) is either stable or unstable. But maybe I misremember.

If that is so, then wouldn't the consistent thing to do with `const fn` be to treat const-unannounced `const fn` as implicitly `rustc_const_usntable`,*even if* they are `#[stable]`?

**Comment by ecstatic-morse** [comment]
> turns out visibility is taken into account by the normal stability system, that is skip_stability_check_due_to_privacy. So maybe the const-stability system should do the same.

This seems right. I mentioned this issue in #76618, but tweaking these rules can be done separately. I'll need to do a bit of refactoring to incorporate them into `check_consts` is all.

**Comment by dtolnay** [comment]
Fixed by the combination of #131349 + #97177.
